packages/lib/src/ecdsa.ts

```diff
@@ -0,0 +1,70 @@
+const ECSignature = require("elliptic/lib/elliptic/ec/signature");
+const BN = require("bn.js");
+import { WeierstrassPoint, babyjubjub } from "./babyJubjub";
+import { Signature } from "./types";
+import { publicKeyFromString } from "./utils";
+
+export const privateKeyToPublicKey = (privKey: bigint): WeierstrassPoint => {
+  const pubKeyPoint = babyjubjub.ec.g.mul(privKey.toString(16));
+  return WeierstrassPoint.fromEllipticPoint(pubKeyPoint);
+};
+
+export const verifyEcdsaSignature = (
+  sig: Signature,
+  msgHash: bigint,
+  pubKey: WeierstrassPoint
+): boolean => {
+  const ecSignature = new ECSignature({
+    r: sig.r.toString(16),
+    s: sig.s.toString(16),
+  });
+  const pubKeyPoint = babyjubjub.ec.curve.point(
+    pubKey.x.toString(16),
+    pubKey.y.toString(16)
+  );
+  const ecPubKey = babyjubjub.ec.keyFromPublic(pubKeyPoint);
+
+  return babyjubjub.ec.verify(msgHash, ecSignature, ecPubKey);
+};
+
+export const recoverPubKeyIndexFromSignature = (
+  sig: Signature,
+  msgHash: bigint,
+  pubKeys: string[]
+): number => {
+  const Fb = babyjubjub.Fb;
+  const Fs = babyjubjub.Fs;
+
+  const pubKeyEdwardsList = pubKeys.map((pubKey) => {
+    const pubKeyWeierstrass = publicKeyFromString(pubKey);
+    return pubKeyWeierstrass.toEdwards();
+  });
+
+  for (let i = 0; i < babyjubjub.cofactor; i++) {
+    for (const parity of [0, 1]) {
+      const r = Fb.add(sig.r, Fb.mul(BigInt(i), Fs.p));
+      const rInv = Fs.inv(r);
+      let R;
+      try {
+        R = babyjubjub.ec.curve.pointFromX(new BN(r.toString(16), 16), parity);
```

Comment:
When does this throw an exception? When the y coordinate is 0?

Reply:
when R is not recoverable from that x coordinate. Will add a comment